crosscluster/producer: flush at lower threshold if consumer is ready

msbutler opened the pull request:
This patch allows the event stream to flush at a lower buffer threshold if the consumer is ready for more work. If a consumer is waiting around for more work, the producer should give them more work to keep consumer busy.

If the consumer is not ready, the event stream will buffer to a higher threshold, preventing a flush from blocking rangefeeds, inducing catchup scans.

Epic: none

Release note: none

msbutler committed:
crosscluster/producer: flush at lower threshold if consumer is ready

This patch allows the event stream to flush at a lower buffer threshold if the
consumer is ready for more work. If a consumer is waiting around for more work,
the producer should give them more work to keep consumer busy.

If the consumer is not ready, the event stream will buffer to a higher
threshold, preventing a flush from blocking rangefeeds, inducing catchup scans.

Epic: none

Release note: none

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/133894)
<!-- Reviewable:end -->

dt commented:
effect of larger buffer size observed here (easier spike is 1mb, later is 32):
<img width="598" alt="Screenshot 2024-10-30 at 17 07 50" src="https://github.com/user-attachments/assets/99c7f932-a8de-4ea2-8242-9e7dee254742">
<img width="559" alt="Screenshot 2024-10-30 at 17 07 43" src="https://github.com/user-attachments/assets/51734c0c-7c6c-493f-8fb1-a3e06411c3ba">

msbutler committed:
crosscluster/logical: bump stream batch size from 1mb to 16mb

By bumping the stream event batch size, fewer flush calls will be made by the
producer, which reduces the chance that a flush call will block rangefeed
progress, preventing catchup scan.

Epic: none

Release note: none

msbutler commented:
tftr!

bors r=dt
